### PR TITLE
Automatic Source Retrieval

### PR DESCRIPTION
**Pull request by CryogenicallyPreservedWombat**
Automatically downloading articles and saving them to CSVs is a useful stepping stone in automating the process of downloading and classifying governmental news releases.

### CODE REVIEW

**Comment by CryogenicallyPreservedWombat** [comment]
There appear to be some issues with source scraping now that it has been implemented for arbitrary time windows. Should not merge until resolved.

**Comment by CryogenicallyPreservedWombat** [comment]
Changes have been pushed to the branch; there shouldn't be any more issues with source scraping.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: Since these functions (eg `_load_newfoundland()`) are all internal you could put the start_date > end_date check in the `load_province()` function and then just assume that all the internal functions will be passed the correct dates. This would be a bit more concise.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: I think you can do this in one line with `full_df = pd.concat(province_dict.values())`

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py:915]: Great documentation on all these functions!

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: One way to save yourself repeating the `load_province(...)` code is to use the `zip()` function in python, which lets you pair up two different lists:

```
provinces = [ 'alberta', 'british columbia', ... ]
provinces_articles = [ load_province( province, start_date=start_date, end_date=end_date, update_csv=update_csv, verbose=verbose ) for province in provinces ]

provinces_results = dict( zip( provinces, provinces_articles ) )
```

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: This is a bit of a nitpick (#nit) but since the `max(...)` has so much going, I might move it to its own line. A good rule of thumb in python is to keep the line lengths to less than 120 characters if possible.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: So this might be a bit of a corner case, but let's say you load articles up to July 27th, and there is an article titled XYZ released at 8:00 in Ontario. Then on July 30th you re-run, going back to July 27th to pick up any new articles. What if, in the meantime, the province changed the title of the article from XYZ to ABC, even though all the content remained the same, or the posting time?

This would then maintain both the original article, as well as the updated article in the database, correct?

It might be worth thinking a bit about what the best way to drop duplicates is (i.e. maybe just drop duplicate urls?)

I don't know that this has actually happened and we don't need to solve it now, but perhaps leave a comment in the code or a `#todo` that articulates this.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: You can take out the `# Not likely to be relevant...` comment. I think this is a smart thing to cover.

Comment (jajsmith) [reply source_scraping.py]: Also, maybe add a comment about how this is searching in reverse order (i.e. from end_date back to start_date). The code is clear, but using the `-1` parameter of `range()` is pretty uncommon and it might be good to reminder future readers of what they should be looking for here.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py:645]: Does the timezone change at all with daylight savings?

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py:861]: It might be nice to add a (verbose) warning for each province if no new articles are found, or maybe just printing out the number of articles. This could help catch any future changes to website structures if we don't run it for a while and then try again. This probably wouldn't make a difference day-to-day.

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: This one could be moved as well to `load_province()`

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py:466]: Well done figuring all of these out!

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: You can remove these kinds of comments

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: cleanup

File: source_scraping.py
Comment (jajsmith) [inline source_scraping.py]: This is clever!

Have you tried using the frequency parameter in date_range? I believe you could get a similar result with:

```
pd.date_range(start=start_date end=end_date, freq="M")
```

See the documentation on pandas date time support here too, its really interesting: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases

**Review by jajsmith — CHANGES_REQUESTED** [review]
I've added a few comments on making the code more concise, but overall this is fantastic! Great job making the code modular, and abstracting the right functionality. This has gone from a very scrappy notebook into a usable library that will continue to be maintainable and functional for a long time. :) 

The one major request is to tweak the csv writer so that all the newlines in the full_text and title columns are stripped, so that each file has only one line per row. This helps with reading the csvs in other programs like Excel. The way I did this in the main NPI csv is:

```
object_columns = df.dtypes[df.dtypes == 'object'].index.values
df[object_columns] = df[object_columns].replace('\n',' ', regex=True)
df[object_columns] = df[object_columns].replace('\r',' ', regex=True)
df.to_csv('output.csv')
```

Once that change is made, I will approve this.

File: source_scraping.py
Comment (CryogenicallyPreservedWombat) [reply source_scraping.py]: It might be easier to move the block into the `_load_province()` function instead, as `load_function()` chops up the provided interval and passes each chunk to `_load_province()`. Would that be alright?

File: source_scraping.py
Comment (CryogenicallyPreservedWombat) [reply source_scraping.py]: That's much neater, thanks!

File: source_scraping.py
Comment (CryogenicallyPreservedWombat) [reply source_scraping.py:645]: The string, at least, doesn't seem to. I've checked as far back as last year's July, and each date ends in -0600. Reading about Alberta's timezones though, that does seem surprising.

File: source_scraping.py
Comment (CryogenicallyPreservedWombat) [reply source_scraping.py:861]: Would it be useful to measure the lengths of the DataFrames before and after loading and outputting the difference as the amount of articles retrieved?

File: source_scraping.py
Comment (jajsmith) [reply source_scraping.py]: Good idea!

File: source_scraping.py
Comment (CryogenicallyPreservedWombat) [reply source_scraping.py]: I should've known there was a less ugly way to do that... thanks for pointing me in the right direction!

**Comment by CryogenicallyPreservedWombat** [comment]
I've noticed that with `pd.date_range()` replacing the messier alternative I had previously written, the first month of the date range is skipped over. I will work on a workaround, but in the meantime it does not behave properly.

File: source_scraping.py
Comment (jajsmith) [reply source_scraping.py:645]: Thanks for checking!

File: source_scraping.py
Comment (jajsmith) [reply source_scraping.py:861]: That makes sense

**Comment by CryogenicallyPreservedWombat** [comment]
I've done some more testing and all the bugs appear to be resolved. Initially some articles were being duplicated, but that was due to checking for duplicates before stripping new lines. That should be fixed now.

**Comment by jajsmith** [comment]
Changes are great!

**Pull request merged**
